web/docs/understand/operators/transformations/flatten.md

```diff
@@ -0,0 +1,68 @@
+# flatten
+
+Flattens nested data structures.
+
+## Synopsis
+
+```
+flatten [<seperator>]
+```
+
+## Description
+
+The `flatten` operator removes any nested lists or records by merging lists and
+joining nested records with a separator. Flattening removes null values inside
+lists.
```

Comment:
Why is flattening removing nulls? This changes the shape of the data and may lead to incorrect results, e.g., when getting the length of a list.

Reply:
It necessarily has to. The purpose of the flatten operator is to bring data into a shape that makes it usable for formats or downstream tools that do not support nested structures. This includes all combinations of lists and records.

When you flatten a list of a list of a type into a list of the type, then you must remove all nulls you encounter on the way:

E.g., `[[1, null], null,  [2, 3]]` has to be flattened into `[1, 2, 3]`. Do you see another option?

Reply:
I think my main problem is that record and list flattening are done in one shot here. These are quite different use cases. If it'd be possible to toggle them separately, e.g., with `-l` for only lists and `-r` for only records, I think this works.

The idea would be still stepping through lists, e.g., when you have a list of a record. Basically `-r` would only latch onto records, where ever they are.

Reply:
I would argue that this should be use-case driven. For now, the only thing we need is a single `flatten` operation that does both, because that's what we want to have for printing csv (and similar) that is guaranteed to be usable by downstream tools.

If we ever have a concrete use case where we need both, then we could separate them out or add an option to flatten only records or only lists, but right now that seems like a lot of extra effort to me that is not properly justified.